scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -584,29 +593,66 @@ s32 pick_idle_cpu(struct task_struct *p, s32 prev_cpu,
 		return -1;
 	}
 
-	idle_smtmask = scx_bpf_get_idle_smtmask();
+	pref_idle_cpumask = bpf_cpumask_create();
 
-	/*
-	 * If CPU has SMT, any wholly idle CPU is likely a better pick than
-	 * partially idle @prev_cpu.
-	 */
-	if ((cpu = pick_idle_cpu_from(layered_cpumask, prev_cpu,
-				      idle_smtmask)) >= 0)
-		goto out_put;
+	if (disable_topology) {
+		/*
+		 * If CPU has SMT, any wholly idle CPU is likely a better pick than
+		 * partially idle @prev_cpu.
+		 */
+		idle_cpumask = scx_bpf_get_idle_smtmask();
+		if ((cpu = pick_idle_cpu_from(layered_cpumask, prev_cpu,
+					      idle_cpumask)) >= 0)
+			goto out_put;
+	} else {
+		/*
+		 * Try a CPU in the current LLC
+		 */
+		idle_cpumask = scx_bpf_get_idle_cpumask();
+		if (!pref_idle_cpumask || !idle_cpumask) {
+			cpu = -1;
+			goto out_put;
+		}
+		bpf_cpumask_copy(pref_idle_cpumask, idle_cpumask);
+		bpf_cpumask_and(pref_idle_cpumask, cache_cpumask, pref_idle_cpumask);
+		bpf_cpumask_and(pref_idle_cpumask, layer_cpumask, pref_idle_cpumask);
```

Comment:
This is likely okay for now but cpumasks can become rather big and these cpumask manipulations can become expensive. Down the line, it probably would make sense to maintain these masks persistently in each layer.

Reply:
That makes sense and then we'd only have to update the masks when the layers change size.